---
title: "Introduction to Biostrings"
author: "Maruf Ahmed Bhuiyan"
date: "7/25/2020"
output: 
  word_document:
    toc: yes
---
```{r}
setwd("/Users/marufahmedbhuiyan/Desktop/DataCamp/Bioconductor")
```

# Exploring the Zika virus sequence
It's your turn to explore the Zika virus genome, which has been loaded in your workspace as zikaVirus. The genome was downloaded from NCBI and you can apply Biostrings functions to learn more about it.

Start by checking the alphabet of the sequence.

alphabet() # Shows the letters included in the sequence
alphabetFrequency() # Shows the counts per letter
Remember from the video that each alphabet corresponds to a specific biostring container, and each alphabet usually has extra code letters and symbols.

```{r}
# Load packages
library(Biostrings)

# Load the zikavius genome in fasta format, 
# NCBI Reference Sequence: NC_012532.1
zikaVirus <- readDNAStringSet('zika.fasta')

# Check the alphabet of the zikaVirus
alphabet(zikaVirus)

# Check the alphabetFrequency of the zikaVirus
alphabetFrequency(zikaVirus) 


# Check alphabet of the zikaVirus using baseOnly = TRUE
alphabet(zikaVirus, baseOnly = TRUE)
```

# Biostrings containers
Now that you know how to check the alphabet of a sequence, can you check what the container class of the zikaVirus object is?

```{r}
# Check alphabet of the zikaVirus using baseOnly = TRUE
alphabet(zikaVirus, baseOnly = TRUE)

# This is a DNAStringset
```

# Manipulating Biostrings
Using a short sequence (dna_seq) from the zikaVirus object, it is your turn to have a go with the two biological processes of transcription and translation.

In the first two parts of this exercise, you will apply these processes separately. In the last part, you'll apply them in one step.

You'll be using a very small sequence in this exercise, but remember that the power of Biostrings comes to light when manipulating much larger sequences.

The Biostrings package has already been loaded for you. Using zikaVirus, you will translate the first 21 characters into an AAString.

```{r}
# Unlist the set, select the first 21 letters, and assign to dna_seq
dna_seq <- subseq(unlist(zikaVirus), end = 21)
dna_seq

# Transcribe dna_seq into an RNAString object and print it
rna_seq <- RNAString(dna_seq) 
rna_seq

# Translate rna_seq into an AAString object and print it
aa_seq <- translate(rna_seq)
aa_seq

# Transcribe and translate dna_seq into an AAString object and print it
aa_seq <- translate(dna_seq)
aa_seq

# Well done! You were able to get an "AAString" from a "DNAString"!
# That's excellent. Next time, you can use the shortcut because the
# function translate() accepts both a "DNAString" and an "RNAString".
# After this lesson, you might want to go and test the functions you've
# just learned with sets and see how powerful these tools are!
```

# From a set to a single sequence
From the video, you know that sequences can be loaded into R using the function readDNAStringSet(). The zikaVirus has been read into your environment using this function.

It is your turn to convert this set into a single sequence, explore the new sequence, and subset it.

```{r}
# Create zikv with one collated sequence using zikaVirus
zikv <- unlist(zikaVirus)

# Check the length of zikaVirus and zikv
length(zikaVirus)
length(zikv)

# Check the width of zikaVirus
width(zikaVirus)

# Subset zikv to only the first 30 bases
subZikv <- subseq(zikv, end = 30)
subZikv
```

# Subsetting a set
In the previous exercise, you've used subseq() to subset a single sequence. Here, you can try subseq() using a set with 3 sequences. The arguments are the same as before: object, start, and end. The last two should be in the form of a vector, for each of the sequences on the set.

subseq(zikaSet, 
        start = c(20, 40, 2), 
        end = c(50, 45, 22)
     )
What is the width of the sequences after subsetting?

```{r}
zikaset <- DNAStringSet(zikv, 
                        start = c(20, 40, 2), 
                        end = c(50, 45, 22))
zikaset
```

# Common sequence manipulation functions
So far, you've learned the the most common sequence manipulation functions:

reverse()
complement()
reverseComplement()
translate()
In real life, you can manipulate really large sequences using these functions.

However, for you to visually see the value and the results of these functions in this exercise, you will use a small subset. The object zikv, which you have previously subsetted from the Zika genome, has only 30 bases.

```{r}
subZikv

reverse(subZikv)
complement(subZikv)
reverseComplement(subZikv)
translate(subZikv)
```
# Searching for a pattern
Let's find some occurrences of a pattern in the zikaVirus set using vmatchPattern(). Then, let's try the same pattern search using matchPattern() with a single sequence, zikv.

# For Sets
vmatchPattern(pattern = "ACATGGGCCTACCATGGGAG", 
              subject = zikaVirus, max.mismatch = 1)
# For single sequences
matchPattern(pattern = "ACATGGGCCTACCATGGGAG", 
              subject = zikv, max.mismatch = 1)
Both functions should find the same number of occurrences, but you will notice a different output. How many matches do we get when running each pattern search individually?

```{r}
# For Sets
vmatchPattern(pattern = "ACATGGGCCTACCATGGGAG", 
              subject = zikaVirus, max.mismatch = 1)

# For single sequences
matchPattern(pattern = "ACATGGGCCTACCATGGGAG", 
             subject = zikv, max.mismatch = 1)
```

# Finding Palindromes
It is your turn to find palindromic sequences using the zikv1000 sequence. Remember, findPalindromes() searches in a single sequence and does not work with a set.

```{r}
zikv1000 <- subseq(zikv, end = 1000)
zikv1000

# Find palindromes in zikv1000
findPalindromes(zikv1000)

```

# Finding a conserved region within six frames
Congratulations on getting this far!

Now you will be able to look for the NS5 protein sequence in the Zika virus sequence. The NS5 is a very conserved virus protein. It was downloaded and loaded for you from Uniprot.

The Zika virus DNA sequence has been transcribed into an RNAStringSet, called rnaframesZikaSet. The set has six reading frames (one per sequence) for you to translate into amino acids. When doing the search, you will set the max.mismatch argument in your call of vcountPattern() to add flexibility to your search.

```{r}
NS5 <- readAAStringSet('NS5.fasta')
NS5

# Reading Frames available from https://www.ncbi.nlm.nih.gov/orffinder/
rnaframesZikaSet <-readAAStringSet('rnaframesZikaSet.fasta')
rnaframesZikaSet

# print the rnaframesZikaSet
rnaframesZikaSet # It's already translated

# translate all 6 reading frames 
# AAzika6F <- translate(rnaframesZikaSet)
# AAzika6F

# Count the matches allowing 15 mistmatches
#vcountPattern(pattern = NS5, subject = rnaframesZikaSet, max.mismatch = 15)

# Select the frame that contains the match
# selectedSet <- AAzika6F[3] 
# selectedSet

# Convert this frame into a single sequence
# selectedSeq <- unlist(selectedSet)
# selectedSeq

```
# Looking for a match
From the previous exercise, you have two objects: selectedSet(a set) and selectedSeq (a single sequence).

You have recently discovered that pattern ns5 is on frame 3 of the AAzika6F. Now, you will discover how the match looks like using selectedSet and selectedSeq.

```{r}
# selectedSeq
# selectedSet

# Use vmatchPattern() with the set
# vmatchPattern(pattern = ns5, subject = selectedSet, max.mismatch = 15)

# Use matchPattern() with the single sequence
# matchPattern(pattern = ns5, subject = selectedSeq, max.mismatch = 15)

# Take your time to see the similarities/differences in the result.
```




# Genetic Code Table from seqinr package

```{r}
# install.packages("seqinr")
library(seqinr)
tablecode()
```

